spec/consensus/proposer-based-timestamp/pbts_001_draft.md

```diff
@@ -0,0 +1,268 @@
+   // add "now_p"
+   proposal ← (validValue_p, now_p) 
+  } else {
+   // add "now_p"
+   proposal ← (getValue(), now_p) 
+  }
+  broadcast ⟨PROPOSAL, h_p, round_p, proposal, validRound_p⟩
+ } else {
+  schedule OnTimeoutPropose(h_p,round_p) to 
+   be executed after timeoutPropose(round_p)
+ }
+}
+```
+
+</td>
+</tr>
+</table>
+
+- Rule on lines 28-35
+
+<table>
+<tr>
+<th>arXiv paper</th>
+<th>Proposer-based time</th>
+</tr>
+
+<tr>
+<td>
+
+```go
+upon timely(⟨PROPOSAL, h_p, round_p, v, vr⟩) 
+ from proposer(h_p, round_p)
+ AND 2f + 1 ⟨PREVOTE, h_p, vr, id(v)⟩ 
+while step_p = propose ∧ (vr ≥ 0 ∧ vr < round_p) do {
+ if valid(v) ∧ (lockedRound_p ≤ vr ∨ lockedValue_p = v) {
+  
+  broadcast ⟨PREVOTE, h_p, round_p, id(v)⟩
+ } else {
+  broadcast ⟨PREVOTE, hp, round_p, nil⟩
+ }
+}
+```
+
+</td>
+
+<td>
+
+```go
+upon timely(⟨PROPOSAL, h_p, round_p, (v, tprop), vr⟩) 
+ from proposer(h_p, round_p) 
+ AND 2f + 1 ⟨PREVOTE, h_p, vr, id(v, tvote)⟩ 
+ while step_p = propose ∧ (vr ≥ 0 ∧ vr < round_p) do {
+  if valid(v) ∧ (lockedRound_p ≤ vr ∨ lockedValue_p = v) {
+   // send hash of v and tprop in PREVOTE message
+   broadcast ⟨PREVOTE, h_p, round_p, id(v, tprop)⟩
+  } else {
+   broadcast ⟨PREVOTE, hp, round_p, nil⟩
+  }
+ }
+```
+
+</td>
+</tr>
+</table>
+
+- Rule on lines 49-54
+
+<table>
+<tr>
+<th>arXiv paper</th>
+<th>Proposer-based time</th>
+</tr>
+
+<tr>
+<td>
+
+```go
+upon ⟨PROPOSAL, h_p, r, v, ∗⟩ from proposer(h_p, r) 
+ AND 2f + 1 ⟨PRECOMMIT, h_p, r, id(v)⟩ 
+ while decisionp[h_p] = nil do {
+  if valid(v) {
+
+   decision_p [h_p] = v
+   h_p ← h_p + 1
+   reset lockedRound_p , lockedValue_p, validRound_p and 
+    validValue_p to initial values and empty message log 
+   StartRound(0)
+  }
+ }
+```
+
+</td>
+
+<td>
+
+```go
+upon ⟨PROPOSAL, h_p, r, (v,t), ∗⟩ from proposer(h_p, r) 
+ AND 2f + 1 ⟨PRECOMMIT, h_p, r, id(v,t)⟩
+ while decisionp[h_p] = nil do {
+  if valid(v) {
+   // decide on time too
+   decision_p [h_p] = (v,t) 
+   h_p ← h_p + 1
+   reset lockedRound_p , lockedValue_p, validRound_p and 
+    validValue_p to initial values and empty message log 
+   StartRound(0)
+  }
+ }
+```
+
+</td>
+</tr>
+</table>
+
+- Other rules are extended in a similar way, or remain unchanged
+
+### Property Overview
+
+#### Safety and Liveness
+
+For safety (Point 1, Point 2, Point 3i) and liveness (Point 4) we need
+the following assumptions:
+
+- There exists a system parameter `PRECISION` such that for any two correct validators `V` and `W`, and at any real-time `t`, their local times `C_V(t)` and `C_W(t)` differ by less than `PRECISION` time units, 
+i.e., `|C_V(t) - C_W(t)| < PRECISION`
+- The message end-to-end delay between a correct proposer and a correct validator (for `PROPOSE` messages) is less than `MSGDELAY`.
+
+#### Relation to Real-Time
+
+For analyzing real-time safety (Point 5), we use a system parameter `ACCURACY`, such that for all real-times `t` and all correct validators `V`, we have `| C_V(t) - t | < ACCURACY`.
```

Comment:
How and where is `ACCURACY` used? Or is it just a metric we can observe within the system?

Reply:
It does not show up in the code, but if there is a value of `ACCURACY` that satisfied `| C_V(t) - t | < ACCURACY`, we can say something about the relation of block-time and real-time. This can be used when observing the chain, e.g., to understand how old a block is.